### PR TITLE
Improvement/support_button&better_performance

### PR DESCRIPTION
**Pull request by CoorFun**
Due to the bug on SAM, need to check with Charles on Monday. Do not merge yet.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,29 +1,47 @@
...
+| **Raspberry Pi 3+**       |  :white_check_mark: | :white_check_mark: | :heavy_minus_sign:   |
+| **ReSpeaker Core V2**     |  :heavy_minus_sign: | :heavy_minus_sign: | :white_check_mark:   |
+
+> Known Issue: the button on the ReSpeaker Core V2 is not yet supported due to its special accessing design
+
+> If you would like to customise the hardware, please refer to [Build Your Own Hardware](https://github.com/snipsco/snips-skill-respeaker/tree/master#build-your-own-hardware) section.
+
 ## Installation
 
 #### By using `sam`:
```
Comment (bjay-wk) [inline README.md:47]: write that sam support only raspberypi 3, 3+ and zero

File: README.md [hunk 2]
```diff
@@ -34,93 +52,222 @@ sam install actions -g https://github.com/snipsco/snips-skill-respeaker.git
 
 #### Manually:
```
Comment (bjay-wk) [inline README.md:56]: snips-skill-server should be install too

File: include/button.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * @brief: Create a button from GPIO port
+ *
+ * @param[in] pin: GPIO port number
+ * @param[in] short_callback: callback function for a short press detection
+ * @param[in] long_callback: callback function for a long press detection
+ *
+ * @returns: -1/ Error
+ *            0/ Success
+ */
+int Init_Key(int pin, int lvl, void (*short_callback)(void), void (*long_callback)(void) );
```
Comment (bjay-wk) [inline include/button.h]: first character of function should be lowercase and don't mix camel case

File: include/button.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @brief: Release the GPIO port which is used for the button and kill the refreshing process
+ *
+ * @returns: -1/ Error
+ *            0/ Success
+ */
+int Destroy_Key(void);
```
Comment (bjay-wk) [inline include/button.h]: first character of function should be lowercase and don't mix camel case

File: include/common.h [hunk 5]
```diff
@@ -9,6 +9,12 @@
 #include <string.h>
 #include <fcntl.h>
 #include <pthread.h>
+#include <signal.h>
```
Comment (bjay-wk) [inline include/common.h]: for includes try to put it in alphabetical order

File: include/gpio_rw.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#define DIRECTION_MAX 35
+
+/**
+ * @brief: export a GPIO port
+ *
+ * @param[in] pin bcm number of the pin
+ *
+ * @returns \-1 failded or \0 successful
+ */
+int GPIO_export(int pin);
```
Comment (bjay-wk) [inline include/gpio_rw.h]: first character of function should be lowercase

File: include/gpio_rw.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @brief: unexport a GPIO port
+ *
+ * @param[in] pin bcm number of the pin
+ *
+ * @returns \-1 failded or \0 successful
+ */
+int GPIO_unexport(int pin);
```
Comment (bjay-wk) [inline include/gpio_rw.h]: same here

File: include/gpio_rw.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @brief: set direction
+ *
+ * @param[in] pin bcm number of the pin
+ * @param[in] dir direction \0 In, \1 Out
+ *
+ * @returns \-1 failded or \0 successful
+ */
+int GPIO_direction(int pin, int dir);
```
Comment (bjay-wk) [inline include/gpio_rw.h]: same here

File: include/gpio_rw.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @brief: read a value from gpio pin
+ *
+ * @param[in] pin bcm number of the pin
+ *
+ * @returns input on this gpio port
+ */
+int GPIO_read(int pin);
```
Comment (bjay-wk) [inline include/gpio_rw.h]: same here

File: include/gpio_rw.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @brief: export a GPIO port
+ *
+ * @param[in] pin bcm number of the pin
+ * @param[in] value \0 Low, \1 High
+ *
+ * @returns \-1 failded or \0 successful
+ */
+int GPIO_write(int pin, int value);
```
Comment (bjay-wk) [inline include/gpio_rw.h]: same here

File: include/load_sw.h [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#ifndef __LOAD_SW_H__
+#define __LOAD_SW_H__
+#include "common.h"
+
+int load_sw_spec(void);
+void debug_run_para_dump(void);
+void dump_running_info(void);
```
Comment (bjay-wk) [inline include/load_sw.h:8]: need one empty line

File: src/button.c [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    time_t curr_time;
+    time(&curr_time);
+    uint8_t pattern_press = ( !button_val << 1) | button_val;
+    uint8_t pattern_release = ( button_val << 1) | !button_val;
+    uint8_t pattern_hold = ( button_val << 1 ) | button_val;
+
+    time_duration = curr_time - start_time;
+    uint8_t res = (last_raw << 1) | (current_raw);
+
+    if( res == pattern_press ){
```
Comment (bjay-wk) [inline src/button.c]: chose one coding style for if then else
and there should not have space after "(" and before ")"

File: src/button.c [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+
+    last_raw = current_raw;
+}
+
+/**
+ * @brief: Read a GPIO key input
+ */
+static void Key_Scan(void){
```
Comment (bjay-wk) [inline src/button.c]: first character of function should be lowercase and don't mix camel case

File: src/button.c [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        if ( button_val == current_raw ) {
+            Key_Status_Handler();
+        }
+    }else{
+        current_raw = !button_val;
+        Key_Status_Handler();
+    }
+}
+
+static void* Key_Status_Observer(void* mqtt_client){
```
Comment (bjay-wk) [inline src/button.c]: same here

File: src/gpio_rw.c [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+int GPIO_export(int pin){
+#define BUFFER_MAX 3
```
Comment (bjay-wk) [inline src/gpio_rw.c]: #define should not be here

File: src/gpio_rw.c [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+	bytes_written = snprintf(buffer, BUFFER_MAX, "%d", pin);
+	write(fd, buffer, bytes_written);
+	close(fd);
+	return(0);
+}
+
+int GPIO_direction(int pin, int dir){
+  	static const char s_directions_str[]  = "in\0out";
+
+ #define DIRECTION_MAX 35
```
Comment (bjay-wk) [inline src/gpio_rw.c]: same here

File: src/gpio_rw.c [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        verbose(V_NORMAL, stderr, BLUE"[%s]"NONE" Failed to set direction!", __FUNCTION__);
+  		return(-1);
+  	}
+
+  	close(fd);
+  	return(0);
+}
+
+int GPIO_read(int pin){
+#define VALUE_MAX 30
```
Comment (bjay-wk) [inline src/gpio_rw.c]: same here

File: src/gpio_rw.c [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+  	char path[VALUE_MAX];
+  	char value_str[3];
+  	int fd;
+
+  	snprintf(path, VALUE_MAX, "/sys/class/gpio/gpio%d/value", pin);
```
Comment (bjay-wk) [inline src/gpio_rw.c]: may be check return value

File: src/gpio_rw.c [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  	}
+
+  	if (-1 == read(fd, value_str, 3)) {
+        verbose(V_NORMAL, stderr, BLUE"[%s]"NONE" Failed to read value!", __FUNCTION__);
+  		return(-1);
+  	}
+
+  	close(fd);
+
+  	return(atoi(value_str));
```
Comment (bjay-wk) [inline src/gpio_rw.c]: use instead strtol()

File: src/load_hw.c [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+#include "cJSON.h"
+#include "gpio_rw.h"
+#include "cAPA102.h"
+#include "verbose.h"
+
+extern SNIPS_RUN_PARA RUN_PARA;
+
+static int get_cJOSN_int(cJSON *spec, const char *key_name){
+    const cJSON *res = cJSON_GetObjectItemCaseSensitive(spec, key_name);
+    if (res != NULL && cJSON_IsNumber(res)) {
```
Comment (bjay-wk) [inline src/load_hw.c]: just use if (res && ...)

File: src/load_hw.c [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+ *
+ * @returns \-1 error or number of loaded bytes
+ */
+static int load_json_file(const char *_DIR, char *json_file_buffer){
+    FILE * pFILE;
+    char temp;
+    int count = 0;
+    verbose(VVV_DEBUG, stdout, BLUE"[%s]"NONE" Opening file: %s ", __FUNCTION__, _DIR);
+    pFILE = fopen(_DIR, "r");
+    if (NULL == pFILE){
```
Comment (bjay-wk) [inline src/load_hw.c]: use if(!pFILE)

File: src/load_hw.c [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+    if(-1 == load_json_file(dir_buffer, buffer))
+        return -1;
+
+    hw_spec_model = cJSON_Parse(buffer);
+    if ( NULL == hw_spec_model){
+        verbose(V_NORMAL, stderr, "No hardware specification found for model: %s", RUN_PARA.hardware_model);
+        return -1;
+    }
+    if ( -1 == load_LEDs_spec(hw_spec_model) )
+        return -1;
```
Comment (bjay-wk) [inline src/load_hw.c]: you should delete hw_spec_model before the return

File: src/load_sw.c [hunk 23]
```diff
@@ -0,0 +1,223 @@
+#include "load_sw.h"
+#include "cCONFIG.h"
+#include "verbose.h"
+
+extern SNIPS_RUN_PARA RUN_PARA;
+
+static uint32_t text_to_colour(const char* cTxt) {
+    if (strlen(cTxt) != 0) {
```
Comment (bjay-wk) [inline src/load_sw.c]: not need of != 0

File: src/load_sw.c [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+        if (strcmp(cTxt, "red") == 0) {
```
Comment (bjay-wk) [inline src/load_sw.c]: if (!strcmp(cTxt, "red"))

**Review by bjay-wk — CHANGES_REQUESTED** [review]

File: src/gpio_rw.c (hunk 18 above)
Comment (CoorFun) [reply src/gpio_rw.c]: No need, pin value will be valid otherwise it will return before

File: src/gpio_rw.c (hunk 19 above)
Comment (CoorFun) [reply src/gpio_rw.c]: No need, the value from `/gpio_x/value` will be either 0 or 1

**Pull request merged**